module2/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,67 @@
+#include "compression.hpp"
+
+#include <iostream>
+#include <algorithm>
+
+std::vector<std::pair<uint8_t, uint8_t>> compressGrayscale(std::array<std::array<uint8_t, width>, height>& getGray) {
+
+    std::vector<std::pair<uint8_t, uint8_t>> smallMap;
+
+    std::transform(getGray.begin(), getGray.end(), getGray.begin(),
+
+                  [&smallMap](std::array<uint8_t,width> everyRow) {
+
+                    auto firstIt = everyRow.begin();
+
+                    while (firstIt != everyRow.end()) {
+                        auto secondIt = std::find_if_not(firstIt, everyRow.end(), [firstIt](int x){ return x == *firstIt; });
+                        smallMap.emplace_back(std::make_pair(*firstIt,std::distance(firstIt, secondIt)));
```

Comment:
Did you know that you can make pair without calling std::make_pair?
Like this:
smallMap.emplace_back(*firstIt, std::distance(firstIt, secondIt) );

Reply:
Exactly, what empalce_back is. When you use make_pair you create temporary object, so use push_back. Emplace_back call directly constructor, so you can use `smallMap.emplace_back(*firstIt, std::distance(firstIt, secondIt) );`

Reply:
in this case is better to use emplace_back or push_back? I still don't understand differences